udata/core/dataset/models.py

```diff
@@ -521,8 +521,13 @@ def save(self, *args, **kwargs):
         self.dataset.save(*args, **kwargs)
 
 
+def validate_badge(value):
+    if value not in Dataset.__badges__.keys():
+        raise db.ValidationError("Unknown badge type")
+
+
 class DatasetBadge(Badge):
-    kind = db.StringField(required=True, choices=list(BADGES.keys()))
+    kind = db.StringField(required=True, validation=validate_badge)
```

Comment:
using `choices` makes it use a dropdown in swagger, which is convenient, but I don't know any existing way to have a dynamic list of choices... so I guess we have to let go of that convenience ;)

Reply:
I've actually tried keeping `choices` with a lambda to compute the value dynamically but I could not make it work sadely

Reply:
Yep, I tried also in the past, and couldn't make it work. There doesn't seem to be a way to do it, other than hacking/monkeypatching the lib itself, which seemed a little bit too much at the time.

I'll have a deeper look in the future if it's prioritized 👍